kafka_helpers/consumer.go

```diff
@@ -0,0 +1,50 @@
+package kafka_helpers
+
+import (
+	"context"
+	"fmt"
+	"github.com/segmentio/kafka-go"
+	"internship_project/elasticsearch_helpers"
+	"log"
+	"strings"
+)
+
+type KafkaConsumer struct {
+	Reader   *kafka.Reader
+	EsClient elasticsearch_helpers.ElasticsearchClient
+}
+
+func (consumer *KafkaConsumer) Consume() {
+	fmt.Println("KafkaConsumer is ready to consume new messages.")
+	for {
+
+		m, err := consumer.Reader.FetchMessage(context.Background())
+		if err != nil {
+			log.Printf("Error while fetching message")
+		}
+
+		fmt.Printf("message at offset %d: %s = %s\n", m.Offset, string(m.Key), string(m.Value))
+
+		s := strings.Split(string(m.Value), " ")
+
+		if len(s) == 3 {
+			if s[0] == "CREATED" || s[0] == "UPDATED" {
+				go consumer.EsClient.IndexDocument(string(m.Key), s[2])
```

Comment:
There are two problems with how messages are handled here:
1. There is no check if there was an error or not. If an error occurs, the index is committed and data is lost.
2. Stream processing should be single-threaded because if you start a goroutine and it fails, the index was already committed in the main thread. This means that you lost data by skipping the message. You should eliminate the goroutine call here. Kafka partitions are the way to go if you want to parallelize processing.

Reply:
We've removed the goroutine call, it is all done in main thread now.